content/post/performance-metrics-overview.md

```diff
@@ -0,0 +1,111 @@
+
+- low cost of both time and space compared to other metric capturing techniques
+- histograms can be easily aggregated, necessary when not all metrics are stored in a single data structure
+- Allows us to answer: is the 99th percentile latency below this time value?
+
+## Minimizing the Metrics Capturing Overhead
+
+We know what metrics information we need to capture, and we knew how we want to capture it. The next step is to determine how much of this workload is required to be handled by Wallaroo. We needed to aggregate histograms to get a complete picture of components but was it worth the resource cost to process this online in Wallaroo? We determined it wasn't. An external system could take these histograms and perform the aggregations needed without impinging on the resources required by Wallaroo. This means freeing up CPU and memory resources that we'd be using if Wallaroo were also responsible for the aggregation of its metrics. Since metrics would have to be offloaded from Wallaroo in some fashion eventually, we decided to do this as early as possible.
+
+### Push vs. Poll
+
+When deciding how to offload the metrics information from Wallaroo to an external system, we have two options: push or poll. We ended up making the metrics system push-based for several reasons. Using Steps as an example here, if we were polling each for its metrics, we'd be wasting CPU resources if a Step had not completed any work. If we have 300 Steps and only 50 are handling the workload, polling and sending empty stats for the remaining 250 could turn our outgoing connection to the external metrics processing system into a bottleneck. By pushing only from Steps which are completing work, we save on CPU resources and minimize the potential of the metrics receiver bottlenecking from an overload of incoming data.
+
+In the end, it was a combination of several design choices that allowed us to capture the metrics we wanted without greatly impacting the performance of Wallaroo. Maintaining our high-throughput and low-latency principles while capturing metrics was not the most straightforward task, but we’re happy with what we’ve been able to achieve so far.
```

Comment:
This paragraph doesn't seem to fit the `Push vs. Poll` heading. Maybe it needs it's own heading?